File: ring-android/libringclient/src/main/java/net/jami/model/Media.java
package net.jami.model;

import java.util.HashMap;
import java.util.Map;

public class Media {

    private static final String SOURCE_KEY = "SOURCE";
    private static final String MEDIA_TYPE_KEY = "MEDIA_TYPE";
    private static final String LABEL_KEY = "LABEL";
    private static final String ENABLED_KEY = "ENABLED";
    private static final String ON_HOLD_KEY = "ON_HOLD";
    private static final String MUTED_KEY = "MUTED";

    public enum MediaType {
        MEDIA_TYPE_AUDIO,
        MEDIA_TYPE_VIDEO;

        static MediaType parseMediaType(String mediaType) {
            for (MediaType mt : values()) {
                if (mt.name().contains(mediaType)) {
                    return mt;
                }
            }
            return null;
        }

        static String getMediaTypeString(MediaType mediaType) {
            if (mediaType == null) return "NULL";
            return mediaType.name();
        }
    }

    private final String source;
    private final MediaType mediaType;
    private final String label;
    private final boolean enabled;
    private final boolean onHold;
    private boolean muted;

    public Media(String source, MediaType mediaType, String label, boolean enabled, boolean onHold, boolean muted) {
        this.source = source;
        this.mediaType = mediaType;
        this.label = label;
        this.enabled = enabled;
        this.onHold = onHold;
        this.muted = muted;
    }

    public Media(Map<String, String> mediaMap) {
        this.source = mediaMap.get(SOURCE_KEY);
        this.mediaType = MediaType.parseMediaType(mediaMap.get(MEDIA_TYPE_KEY));
        this.label = mediaMap.get(LABEL_KEY);
        this.enabled = Boolean.parseBoolean(mediaMap.get(ENABLED_KEY));
        this.onHold = Boolean.parseBoolean(mediaMap.get(ON_HOLD_KEY));
        this.muted = Boolean.parseBoolean(mediaMap.get(MUTED_KEY));
    }

    public String getSource() {
        return source;
    }

    public MediaType getMediaType() {
        return mediaType;
    }

    public String getLabel() {
        return label;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public boolean isOnHold() {
        return onHold;
    }

    public boolean isMuted() {
        return muted;
    }

    public void setMuted(boolean muted) {
        this.muted = muted;
    }

    public Map<String, String> toMap() {
        Map<String, String> stringStringMap = new HashMap<>();
        stringStringMap.put(SOURCE_KEY, this.source);
        stringStringMap.put(MEDIA_TYPE_KEY, MediaType.getMediaTypeString(this.mediaType));
        stringStringMap.put(LABEL_KEY, this.label);
        stringStringMap.put(ENABLED_KEY, Boolean.toString(this.enabled));
        stringStringMap.put(ON_HOLD_KEY, Boolean.toString(this.onHold));
        stringStringMap.put(MUTED_KEY, Boolean.toString(this.muted));
        return stringStringMap;
    }
}

File: ring-android/libjamiclient/src/main/kotlin/net/jami/model/Media.kt
package net.jami.model

import net.jami.daemon.StringMap
import java.util.*

data class Media(val source: String?,
                 val mediaType: MediaType?,
                 val label: String?,
                 val isEnabled: Boolean,
                 val isOnHold: Boolean,
                 val isMuted: Boolean
) {
    constructor(mediaMap: Map<String, String>) : this(
            source = mediaMap[SOURCE_KEY],
            mediaType = MediaType.parseMediaType(mediaMap[MEDIA_TYPE_KEY]!!),
            label = mediaMap[LABEL_KEY],
            isEnabled = java.lang.Boolean.parseBoolean(mediaMap[ENABLED_KEY]),
            isOnHold = java.lang.Boolean.parseBoolean(mediaMap[ON_HOLD_KEY]),
            isMuted = java.lang.Boolean.parseBoolean(mediaMap[MUTED_KEY])
    )

    constructor(type: MediaType, label: String) : this(
            source = "",
            mediaType = type,
            label = label,
            isEnabled = true,
            isOnHold = false,
            isMuted = false
    )

    enum class MediaType {
        MEDIA_TYPE_AUDIO,
        MEDIA_TYPE_VIDEO;

        companion object {
            fun parseMediaType(mediaType: String): MediaType? {
                for (mt in values()) {
                    if (mt.name.contains(mediaType)) {
                        return mt
                    }
                }
                return null
            }

            fun getMediaTypeString(mediaType: MediaType?): String {
                return mediaType?.name ?: "NULL"
            }
        }
    }

    fun toMap(): StringMap {
        val map = StringMap()
        if (source != null) map[SOURCE_KEY] = source
        map[MEDIA_TYPE_KEY] = MediaType.getMediaTypeString(mediaType)
        if (label != null) map[LABEL_KEY] = label
        map[ENABLED_KEY] = java.lang.Boolean.toString(isEnabled)
        map[ON_HOLD_KEY] = java.lang.Boolean.toString(isOnHold)
        map[MUTED_KEY] = java.lang.Boolean.toString(isMuted)
        return map
    }

    companion object {
        private const val SOURCE_KEY = "SOURCE"
        const val MEDIA_TYPE_KEY = "MEDIA_TYPE"
        private const val LABEL_KEY = "LABEL"
        private const val ENABLED_KEY = "ENABLED"
        private const val ON_HOLD_KEY = "ON_HOLD"
        const val MUTED_KEY = "MUTED"
        val DEFAULT_AUDIO = Media(MediaType.MEDIA_TYPE_AUDIO, "audio_0")
        val DEFAULT_VIDEO = Media(MediaType.MEDIA_TYPE_VIDEO, "video_0")
    }
}
